src/number_theory/legendre_symbol/mul_character.lean

```diff
@@ -0,0 +1,549 @@
+  ..has_mul }
+
+instance semigroup : semigroup (mul_char R R') :=
+{ mul_assoc := by { intros χ₁ χ₂ χ₃, ext a, repeat {rw [coe_to_fun_mul]}, rw [mul_assoc], },
+  ..has_mul }
+
+noncomputable
+instance monoid : monoid (mul_char R R') :=
+{ ..mul_one_class,
+  ..semigroup }
+
+instance comm_semigroup : comm_semigroup (mul_char R R') :=
+{ mul_comm := by { intros χ₁ χ₂, ext a, repeat {rw [coe_to_fun_mul]}, rw [mul_comm], },
+  ..semigroup }
+
+noncomputable
+instance comm_monoid : comm_monoid (mul_char R R') :=
+{ ..monoid,
+  ..comm_semigroup }
+
+/-- The inverse of a multiplicative character. We define it as `inverse ∘ χ`. -/
+noncomputable
+def inv (χ : mul_char R R') : mul_char R R' :=
+{ map_nonunit' :=
+  begin
+    intros a ha,
+    simp only [monoid_hom.to_fun_eq_coe, monoid_hom.coe_comp, function.comp_app,
+               monoid_with_zero_hom.to_monoid_hom_coe, monoid_with_zero.coe_inverse, coe_coe],
+    rw [map_nonunit _ ha, ring.inverse_zero],
+  end,
+  ..monoid_with_zero.inverse.to_monoid_hom.comp χ.to_monoid_hom }
+
+noncomputable
+instance has_inv : has_inv (mul_char R R') := ⟨inv⟩
+
+/-- The inverse of a multiplicative character `χ`, applied to `a`, is the inverse of `χ a`. -/
+lemma inv_apply_eq_inv (χ : mul_char R R') (a : R) :
+  χ⁻¹ a = ring.inverse (χ a) :=
+eq.refl $ inv χ a
+
+/-- Variant when the target is a field -/
+lemma inv_apply_eq_inv' {R' : Type v} [field R'] (χ : mul_char R R') (a : R) :
+  χ⁻¹ a = (χ a)⁻¹ :=
+(inv_apply_eq_inv χ a).trans $ ring.inverse_eq_inv (χ a)
+
+/-- When the domain has a zero, we can as well take the inverse first. -/
+lemma inv_apply {R : Type u} [comm_monoid_with_zero R] (χ : mul_char R R') (a : R) :
+  χ⁻¹ a = χ (ring.inverse a) :=
+begin
+  by_cases ha : is_unit a,
+  { rw [inv_apply_eq_inv],
+    have h := is_unit.map χ ha,
+    apply_fun ((*) (χ a)) using is_unit.mul_right_injective h,
+    rw [ring.mul_inverse_cancel _ h, ← map_mul, ring.mul_inverse_cancel _ ha, mul_char.map_one], },
+  { revert ha, nontriviality R, intro ha, -- `nontriviality R` by itself doesn't do it
+    rw [map_nonunit _ ha, ring.inverse_non_unit a ha, mul_char.map_zero χ], },
+end
+
+/-- When the domain is a field, we can use the field inverse instead. -/
+lemma inv_apply' {R : Type u} [field R] (χ : mul_char R R') (a : R) : χ⁻¹ a = χ a⁻¹ :=
+(inv_apply χ a).trans $ congr_arg _ (ring.inverse_eq_inv a)
+
+noncomputable
+instance div_inv_monoid : div_inv_monoid (mul_char R R') :=
+{ ..monoid,
+  ..has_inv }
+
+/-- The product of a character with its inverse is the trivial character. -/
+@[simp]
+lemma inv_mul (χ : mul_char R R') : χ⁻¹ * χ = 1 :=
+begin
+  ext x hx,
+  rw [coe_to_fun_mul, pi.mul_apply, inv_apply_eq_inv,
+      ring.inverse_mul_cancel _ (is_unit.map _ hx), trivial_unit hx],
+end
+
+noncomputable
+instance group : group (mul_char R R') :=
+{ mul_left_inv := inv_mul,
+  ..div_inv_monoid }
+
+noncomputable
+instance comm_group : comm_group (mul_char R R') :=
+{ ..group,
+  ..comm_monoid }
```

Comment:
```suggestion
noncomputable
instance comm_group : comm_group (mul_char R R') :=
{ mul_left_inv := inv_mul,
  ..div_inv_monoid,
  ..comm_monoid }
```
Again I think it might be better to just make the most important instance?